channeldb/migration14/migration.go

```diff
@@ -0,0 +1,170 @@
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+)
+
+// MigrateDatabaseWireMessages performs a migration in all areas that we
+// currently store wire messages withotu length prefixes. This includes the
+// CommitDiff struct, LogUpdates, and also the networkResult struct as well.
+func MigrateDatabaseWireMessages(tx kvdb.RwTx) error {
+	openChanBucket := tx.ReadWriteBucket(openChannelBucket)
+	if openChanBucket == nil {
+		return nil
+	}
+
+	// The migration will proceed in two phases: we'll need to update any
+	// pending commit diffs and unsigned acked updates for all open
+	// channels in the first phsae. For the second phase, we'll need to
+	// update all the current stored network results for payments in the
+	// switch.
+	//
+	// In this phase, we'll migrate the open channel data.
+	err := openChanBucket.ForEach(func(nodePub, v []byte) error {
+		// Ensure that this is a key the same size as a pubkey,
+		// and also that it leads directly to a bucket.
+		if len(nodePub) != 33 || v != nil {
+			return nil
+		}
+
+		nodeChanBucket := openChanBucket.NestedReadWriteBucket(nodePub)
+		if nodeChanBucket == nil {
+			return fmt.Errorf("no bucket for node %x", nodePub)
+		}
+
+		// The next layer down is all the chains that this node
+		// has channels on with us.
+		return nodeChanBucket.ForEach(func(chainHash, v []byte) error {
+			// If there's a value, it's not a bucket so
+			// ignore it.
+			if v != nil {
+				return nil
+			}
+
+			chainBucket := nodeChanBucket.NestedReadWriteBucket(
+				chainHash,
+			)
+			if chainBucket == nil {
+				return fmt.Errorf("unable to read "+
+					"bucket for chain=%x", chainHash)
+			}
+
+			return chainBucket.ForEach(func(chanPoint, v []byte) error {
+				// If there's a value, it's not a bucket so
+				// ignore it.
+				if v != nil {
+					return nil
+				}
+
+				// Once we've found a valid channel bucket,
+				// we'll extract it from the node's chain
+				// bucket.
+				chanBucket := chainBucket.NestedReadWriteBucket(
+					chanPoint,
+				)
+				if chanBucket == nil {
+					return fmt.Errorf("not bucket for "+
+						"channel: %x", chanPoint)
+				}
+
+				// At this point, we have the channel bucket
+				// now, so we'll check to see if this channel
+				// has a pending commitment or not. If it
+				// doesn't, then we can go on as there's
+				// nothing to migrate.
+				commitDiffBytes := chanBucket.Get(commitDiffKey)
+				if commitDiffBytes == nil {
+					return nil
+				}
+
+				// Now that we have the commit diff in the
+				// _old_ encoding, we'll write it back to disk
+				// using the new encoding which has a length
+				// prefix in front of the CommitSig.
+				commitDiff, err := deserializeCommitDiffLegacy(
+					bytes.NewReader(commitDiffBytes),
+				)
+				if err != nil {
+					return err
+				}
+
+				var b bytes.Buffer
+				err = serializeCommitDiff(&b, commitDiff)
+				if err != nil {
+					return err
+				}
+
+				err = chanBucket.Put(
```

Comment:
We should be careful here, as the bbolt docs state that 
> ForEach executes a function for each key/value pair in a bucket. [...] The provided function must not modify the bucket; this will result in undefined behavior.

Not sure if modifying a subbucket of the bucket counts as modifyinng the bucket itself.

Reply:
Hmm yeah I had that same thought here when writing this section myself...updating to just do multiple passes so we're on the safe side.